server/lib/matchmaking/matchmaking-service.ts

```diff
@@ -575,74 +570,115 @@ export class MatchmakingApi {
+    clientSockets.subscribe(
+      MatchmakingService.getClientPath(clientSockets),
+      undefined,
+      this.removeClientFromMatchmaking,
+    )
   }
 
-  @Api('/cancel')
-  async cancel(data: Map<string, any>, next: NextFunc) {
-    const user = this.getUser(data)
-    const client = activityRegistry.getClientForUser(user.userId)
-    if (!client || !this.queueEntries.has(user.name)) {
-      throw new errors.Conflict('user does not have an active matchmaking queue')
+  async cancel(userId: number) {
+    const userSockets = this.getUserSockets(userId)
+    const clientSockets = activityRegistry.getClientForUser(userSockets.userId)
+    if (!clientSockets || !this.queueEntries.has(userSockets.name)) {
+      throw new MatchmakingServiceError(
+        MatchmakingServiceErrorCode.NotInQueue,
+        'User does not have an active matchmaking queue',
+      )
     }
 
-    this.handleLeave(client)
+    this.removeClientFromMatchmaking(clientSockets)
   }
 
-  @Api('/accept')
-  async accept(data: Map<string, any>, next: NextFunc) {
-    const client = this.getClient(data)
-    if (!this.acceptor.registerAccept(client)) {
-      throw new errors.NotFound('no active match found')
+  async accept(userId: number) {
+    const userSockets = this.getUserSockets(userId)
+    const clientSockets = activityRegistry.getClientForUser(userSockets.userId)
+    if (!clientSockets || !this.acceptor.registerAccept(clientSockets)) {
+      throw new MatchmakingServiceError(
+        MatchmakingServiceErrorCode.NoActiveMatch,
+        'No active match found',
+      )
     }
   }
 
-  getUser(data: Map<string, any>) {
-    const user = this.userSockets.getBySocket(data.get('client'))
-    if (!user) throw new errors.Unauthorized('authorization required')
-    return user
+  private removeClientFromMatchmaking(client: ClientSocketsGroup) {
+    // NOTE(2Pac): Client can leave, i.e. disconnect, during the queueing process, during the
+    // loading process, or even during the game process.
+    const entry = this.queueEntries.get(client.name)
+    // Means the client disconnected during the queueing process
+    if (entry) {
+      this.queueEntries = this.queueEntries.delete(client.name)
+      this.matchmakers.get(entry.type)!.removeFromQueue(entry.username)
+      this.acceptor.registerDisconnect(client)
+    }
+
+    // Means the client disconnected during the loading process
+    if (this.clientTimers.has(client.name)) {
+      const { mapSelectionTimer, countdownTimer, cancelToken } = this.clientTimers.get(client.name)!
+      if (countdownTimer) {
+        countdownTimer.reject(new Error('Countdown cancelled'))
+      }
+      if (mapSelectionTimer) {
+        mapSelectionTimer.reject(new Error('Map selection cancelled'))
+      }
+
+      cancelToken.cancel()
+
+      this.clientTimers = this.clientTimers.delete(client.name)
+    }
+
+    this.unregisterActivity(client)
+  }
+
+  private getUserSockets(userId: number): UserSocketsGroup {
```

Comment:
Not something you caused, but it might be good to name these with `OrFail` as a postfix. We write these different ways sometimes, and it's really not obvious in the code using them whether or not they throw if the retrieval fails (and we've caused problems before by adding this sort of call somewhere, not realizing it will break when the user is offline, etc.). So I think e.g. `getUserSocketsOrFail` makes that a lot clearer, and makes us less likely to introduce such bugs

Reply:
Well, I'd say that's a good practice to have going forward as immediately as I renamed this method I noticed I'm calling it when unregistering an activity, which shouldn't really throw errors if it fails ^^.